Tools/clinic/clinic.py

```diff
@@ -116,8 +116,28 @@ def text_accumulator():
     text, append, output = _text_accumulator()
     return TextAccumulator(append, output)
 
-
-def warn_or_fail(fail=False, *args, filename=None, line_number=None):
+@overload
+def warn_or_fail(
+    *args: object,
+    fail: Literal[True],
+    filename: str | None = None,
+    line_number: int | None = None,
+) -> NoReturn: ...
```

Comment:
This is neat

Reply:
We might want to add `warn_unreachable = True` to https://github.com/python/cpython/blob/main/Tools/clinic/mypy.ini if we do this. Otherwise, mypy might start inferring blocks of code as unreachable, and would stop checking those blocks of code without telling us about it

Reply:
Definitely sounds like something we want in our config.